File: katas/7 kyu/printer-errors/solution.py
from turtle import width
import codewars_test as test
# -------------------------------- my solution ------------------------------- #
def printer_error(s):
  count = 0
  for x in s.upper():
    if ord(x) not in range(65,78):
      count += 1 
  return f'{count}/{len(s)}'
# ------------------------------ clever solution ----------------------------- #
from re import sub
def printer_error(s):
  return "{}/{}".format(len(sub("[a-m]",'',s)),len(s))

def printer_error(s):
  return "{}/{}".format(len([x for x in s if x not in "abcdefghijklm"]), len(s))
# ----------------------------------- test ----------------------------------- #
s="aaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbmmmmmmmmmmmmmmmmmmmxyz"
test.assert_equals(printer_error(s), "3/56")
s = "kkkwwwaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbmmmmmmmmmmmmmmmmmmmxyz"
test.assert_equals(printer_error(s), "6/60")
s = "kkkwwwaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbmmmmmmmmmmmmmmmmmmmxyzuuuuu"
test.assert_equals(printer_error(s) , "11/65")

File: katas/7 kyu/printer-errors/solution.php
<!-- printer-errors -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function printerError($s) {
  $sample = 'abcdefghijklm';
  $arr = str_split($s);
  $error = 0;
  foreach ($arr as $chr){
    if (strpos($sample , $chr) === false){
      $error ++ ;
    }
  }
  return $error . "/" . strlen($s);
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
function printerError($s) {
  return strlen(preg_replace('/[a-m]/i', '', $s)) . '/' . strlen($s);
}

function printerError($s) {
  return preg_match_all('/[^a-m]/', $s).'/'.strlen($s);
}
/* ---------------------------------- TEST ---------------------------------- */
class PrinterErrorTestCases extends TestCase {
  private function revTest($actual, $expected) {
      $this->assertEquals($expected, $actual);
  }
  public function testBasics() {        
      $s="aaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbmmmmmmmmmmmmmmmmmmmxyz";
      $this->revTest(printerError($s), "3/56");
      $s = "kkkwwwaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbmmmmmmmmmmmmmmmmmmmxyz";
      $this->revTest(printerError($s), "6/60");
      $s = "kkkwwwaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbmmmmmmmmmmmmmmmmmmmxyzuuuuu";
      $this->revTest(printerError($s) , "11/65");
  }
}
